src/client.ts

```diff
@@ -1,41 +1,51 @@
 // SPDX-License-Identifier: Apache-2.0
 "use strict";
 
-import { Signer } from "ethers";
-import { ethers } from "ethers";
 
-import { Erdstall } from "erdstall";
+import { ErdstallClient } from "erdstall";
 import { Erdstall__factory } from "#erdstall/ledger/backend/contracts";
-import { TxReceipt } from "#erdstall/api/responses";
-import { BalanceProof } from "#erdstall/api/responses";
 import { ClientConfig } from "#erdstall/api/responses";
-import { Transfer, Mint, ExitRequest, TradeOffer, Trade } from "#erdstall/api/transactions";
-import { EnclaveConnection, EnclaveEvent } from "#erdstall/enclave";
-import { LedgerConnection, LedgerAdapter } from "#erdstall/ledger";
-import { ErdstallEvent, isLedgerEvent } from "#erdstall/ledger";
-import { Assets } from "#erdstall/ledger/assets";
-import { Address } from "#erdstall/ledger";
-import { Uint256 } from "#erdstall/api/util";
-import { EventCache, OneShotEventCache, Stages } from "#erdstall/utils";
-
-export const ErrUnitialisedClient = new Error("client unitialised");
-
-export default class Client implements Erdstall {
-	readonly address: Address;
-	private nonce: bigint;
-	private enclaveConn: EnclaveConnection;
-	private erdstallConn?: LedgerConnection;
+import {
+	Enclave,
+	EnclaveWSProvider,
+	EnclaveReader,
+	EnclaveWriter,
+	EnclaveEvent
+} from "#erdstall/enclave";
+import {
+	LedgerWriteConn,
+	LedgerReader,
+	LedgerWriter,
+	Address,
+	ErdstallEvent,
+	isLedgerEvent
+} from "#erdstall/ledger";
+import { EventCache, OneShotEventCache } from "#erdstall/utils";
+import { ethers, Signer } from "ethers";
+
+export class Client implements ErdstallClient {
+	protected enclaveConn: EnclaveReader | EnclaveWriter;
+	protected provider: ethers.providers.Provider | Signer;
+	protected erdstallConn?: LedgerReader | LedgerWriter;
 	private erdstallEventHandlerCache: EventCache<ErdstallEvent>;
 	private erdstallOneShotHandlerCache: OneShotEventCache<ErdstallEvent>;
-	private signer: Signer;
 
-	constructor(address: Address, signer: Signer, encConn: EnclaveConnection) {
-		this.address = address;
-		this.signer = signer;
-		this.nonce = 1n;
-		this.enclaveConn = encConn;
-		this.erdstallEventHandlerCache = new EventCache<ErdstallEvent>();
-		this.erdstallOneShotHandlerCache = new OneShotEventCache<ErdstallEvent>();
+	// Takes (Client) or (Provider | Signer, EnclaveReader | EnclaveWriter | URL).
+	constructor(clientOrLedgerProvider: Client | ethers.providers.Provider | Signer, encConn?: EnclaveReader | EnclaveWriter | URL) {
+		if(clientOrLedgerProvider instanceof Client) {
+			this.provider = clientOrLedgerProvider.provider;
+			this.enclaveConn = clientOrLedgerProvider.enclaveConn;
+			this.erdstallEventHandlerCache = clientOrLedgerProvider.erdstallEventHandlerCache;
+			this.erdstallOneShotHandlerCache = clientOrLedgerProvider.erdstallOneShotHandlerCache;
+		} else {
```

Comment:
Wouldn't it be possible to remove the constructor taking a `Client`? And the `Session` constructor should then receive `(address: Address, singer: Signer, encConn: EnclaveWriter | URL)`. I think you'd also have better type safety this way and less type checks.

IMO you could also remove the `| URL` option from the constructors and then pass the `URL` to the `.dial` statics (could create one for Client).

Reply:
And I think you can also omit `| EnclaveWriter` because it extends a reader. So it should be able to change this ctor to `(provider: ethers.providers.Provider | Signer, encConn: EnclaveReader)`.